### PR TITLE
TC-119 - Create function to dedup suggester lenders

### PR DESCRIPTION
**Pull request by peggles2**
TC-119 This pull request removes the duplicate suggested lenders. So that when its used for autocomplete it does not confuse the user. 

https://brian.ussba.io/api/content/search/suggestions.json?lenderName=Sun%20Tr

### CODE REVIEW

File: src/service/lender-search.js [hunk 1]
```diff
@@ -55,6 +55,31 @@ function buildParams (query, geo) {
   return params
 }
 
+function titleCase (str) {
```
Comment (kscoulter) [inline src/service/lender-search.js:58]: Very good! This is concise and clear function.

File: src/service/lender-search.js [hunk 2]
```diff
@@ -55,6 +55,31 @@ function buildParams (query, geo) {
...
+  var splitStr = str.toLowerCase().split(' ')
+  for (var i = 0; i < splitStr.length; i++) {
+    splitStr[i] = splitStr[i].charAt(0).toUpperCase() + splitStr[i].substring(1)
+  }
+  return splitStr.join(' ')
+}
+
+function dedupSuggester (result) {
```
Comment (kscoulter) [inline src/service/lender-search.js:66]: should be `dedupeSuggestions`, and `result` suggests to me that there will only be one. I would suggest changing it to either `results` or `suggestions`

File: src/service/lender-search.js [hunk 3]
```diff
@@ -55,6 +55,31 @@ function buildParams (query, geo) {
...
+  var i
```
Comment (kscoulter) [inline src/service/lender-search.js:67]: don't use var, it's usually declared in the for loop like this `for(let i = 0; etc.)`

File: src/service/lender-search.js [hunk 4]
```diff
@@ -55,6 +55,31 @@ function buildParams (query, geo) {
...
+  let dedupSuggestions = []
+  let suggestions = result ? result.suggest.suggestions : ''
+
+  for (i = 0; i < suggestions.length; i++) {
```
Comment (kscoulter) [inline src/service/lender-search.js:71]: I would use map for this, it will return an array for you.

```dedupedSuggestions = suggestions.map((suggestion, index)=> {})```

To trim the white space just use `trim()` 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim

File: src/service/lender-search.js [hunk 5]
```diff
@@ -55,6 +55,31 @@ function buildParams (query, geo) {
...
+    dedupSuggestions.push(
+      titleCase(suggestions[i]
+        .suggestion
+        .split(/ -| \|/)[0]
+        .replace(/ atm/gi, '')
+      )
+    )
+  }
+  return [...new Set(dedupSuggestions)]
```
Comment (kscoulter) [inline src/service/lender-search.js:80]: awesome use of spread operator 👍

File: src/service/lender-search.js [hunk 6]
```diff
@@ -65,7 +90,7 @@ async function fetchSuggestions (query) {
     }
     const result = await cloudsearch.runSuggester(params, endpoint) // call the module.exports version for stubbing during testing
```
Comment (kscoulter) [inline src/service/lender-search.js:91]: result is a little vague, I'd use `suggestions` or `data` or `lenderData` to make it more readable.

File: src/service/lender-search.js [hunk 7]
```diff
@@ -55,6 +55,31 @@ function buildParams (query, geo) {
...
+  for (var i = 0; i < splitStr.length; i++) {
+    splitStr[i] = splitStr[i].charAt(0).toUpperCase() + splitStr[i].substring(1)
+  }
+  return splitStr.join(' ')
+}
+
+function dedupSuggester (result) {
+  var i
+  let dedupSuggestions = []
+  let suggestions = result ? result.suggest.suggestions : ''
```
Comment (kscoulter) [inline src/service/lender-search.js:69]: the suggestions are not a string, so turning it into a string if empty is misleading. Should be an empty array. You can also just immediately return if results are empty. 
`if(results.length < 1) return;`

File: src/service/lender-search.test.js [hunk 8]
```diff
@@ -129,13 +168,25 @@ describe('# Lender Search', () => {
...
         lenderSearchRunSuggesterStub.calledWith({
           query: "'Chase Ban'",
           suggester: 'lender_name_suggester',
           size: '10'
         }).should.be.true
       })
+
+      it('should dedupe suggesters as much as possible', async () => {
+        lenderSearchRunSuggesterStub.returns(exampleCloudSuggestResponse)
+        let result = await lenderSearch.fetchSuggestions({ lenderName: 'Capital On' })
```
Comment (kscoulter) [inline src/service/lender-search.test.js:182]: you don't need to call the stub to test the independent function. You should export that function and test it independently of the API endpoint. Make sure to pass empty results and weirdly formatted strings to avoid only testing the happy path.

File: src/service/lender-search.js (hunk 2 above)
Comment (peggles2) [reply src/service/lender-search.js:66]: yeah originally had it at suggestions but then its suggestions.suggest.suggestions which is a mouthful. I am renaming it lenders. easier to read.

File: src/service/lender-search.js [hunk 9]
```diff
@@ -55,6 +55,29 @@ function buildParams (query, geo) {
...
+function titleCase (str) {
+  var splitStr = str.toLowerCase().split(' ')
+  for (var i = 0; i < splitStr.length; i++) {
+    splitStr[i] = splitStr[i].charAt(0).toUpperCase() + splitStr[i].substring(1)
+  }
+  return splitStr.join(' ')
+}
+
+function dedupeLenders (lenders) {
+  let cleanupLenderList = []
```
Comment (kscoulter) [inline src/service/lender-search.js:67]: couple things, you don't have to declare the variable here. Also, funny enough you can use const, because while the contents of the array may change, the variable itself remains an array that you've declared. It's a little counterintuitive.

so `const cleaupLenderLIst  = lenders.map`

File: src/service/lender-search.js [hunk 10]
```diff
@@ -55,6 +55,29 @@ function buildParams (query, geo) {
...
+
+  cleanupLenderList = lenders.map((lender) => {
+    return titleCase(lender
+      .suggestion
+      .split(/ -| \|/)[0]
+      .replace(/ atm/gi, '')
+      .trim()
+    )
+  })
+  return [...new Set(cleanupLenderList)]
```
Comment (kscoulter) [inline src/service/lender-search.js:78]: awesome, very clear and concise!

File: src/service/lender-search.test.js [hunk 11]
```diff
@@ -129,13 +171,58 @@ describe('# Lender Search', () => {
...
+
+      it('should return list of suggesters', async () => {
+        lenderSearchRunSuggesterStub.returns(exampleCloudSuggestResponse)
+        let result = await lenderSearch.fetchSuggestions({ lenderName: 'Capital On' })
+      
+        result.should.eql([ 'Capital One', 'Capital One Bank', 'Capital One Mortgage' ])
+      })
+
+      it('should dedupe list of lenders', () => {
+        lenderSearchRunSuggesterStub.returns(exampleCloudSuggestResponse)
```
Comment (kscoulter) [inline src/service/lender-search.test.js:191]: is this line necessary? I'm not sure what it does

File: src/service/lender-search.test.js [hunk 12]
```diff
@@ -129,13 +171,58 @@ describe('# Lender Search', () => {
...
+
+        lenders[0].should.eq('Bank Of America')
+      })
+
+      it('title cases all lender names', () => {
+        let lenders = lenderSearch.titleCase('bank of america')
+
+        lenders.should.eq('Bank Of America')
+      })
+
```
Comment (kscoulter) [inline src/service/lender-search.test.js:225]: Add another test for empty array to make sure the dedupe function doesn't throw an error when an empty response is passed

File: src/service/lender-search.test.js (hunk 11 above)
Comment (peggles2) [reply src/service/lender-search.test.js:191]: Its testing out dedupeLenders method. So yes I like to keep it.

Comment (peggles2) [reply src/service/lender-search.test.js:191]: Yup not needed taking it out
